pyquil/kraus.py

```diff
@@ -56,3 +58,211 @@ def _create_kraus_pragmas(name, qubit_indices, kraus_ops):
+    :return: A list [k1, k2] of the Kraus operators that parametrize the map.
+    :rtype: list
+    """
+    damping_op = np.sqrt(p) * np.array([[0, 1],
+                                        [0, 0]])
+
+    residual_kraus = np.diag([1, np.sqrt(1 - p)])
+    return [residual_kraus, damping_op]
+
+
+def dephasing_kraus_map(p=0.10):
+    """
+    Generate the Kraus operators corresponding to a dephasing channel.
+
+    :params float p: The one-step dephasing probability.
+    :return: A list [k1, k2] of the Kraus operators that parametrize the map.
+    :rtype: list
+    """
+    return [np.sqrt(1 - p) * np.eye(2), np.sqrt(p) * np.diag([1, -1])]
+
+
+def tensor_kraus_maps(k1, k2):
+    """
+    Generate the Kraus map corresponding to the composition
+    of two maps on different qubits.
+
+    :param list k1: The Kraus operators for the first qubit.
+    :param list k2: The Kraus operators for the second qubit.
+    :return: A list of tensored Kraus operators.
+    """
+    return [np.kron(k1j, k2l) for k1j in k1 for k2l in k2]
+
+
+def combine_kraus_maps(k1, k2):
+    """
+    Generate the Kraus map corresponding to the composition
+    of two maps on the same qubits with k1 being applied to the state
+    after k2.
+
+    :param list k1: The list of Kraus operators that are applied second.
+    :param list k2: The list of Kraus operators that are applied first.
+    :return: A combinatorically generated list of composed Kraus operators.
+    """
+    return [np.dot(k1j, k2l) for k1j in k1 for k2l in k2]
+
+
+def damping_after_dephasing(T1, T2, gate_time):
+    """
+    Generate the Kraus map corresponding to the composition
+    of a dephasing channel followed by an amplitude damping channel.
+
+    :param float T1: The amplitude damping time
+    :param float T2: The dephasing time
+    :param float gate_time: The gate duration.
+    :return: A list of Kraus operators.
+    """
+    damping = damping_kraus_map(p=gate_time / T1)
+    dephasing = dephasing_kraus_map(p=gate_time / T2)
+    return combine_kraus_maps(damping, dephasing)
+
+
+# You can only apply gate-noise to non-parametrized gates,
+# so we need to define placeholders for RX(+/- pi/2)
+SINGLE_Q = {
+    "noisy-x-plus90": expm(-1j * np.pi / 4 * np.array([[0, 1],
+                                                       [1, 0]])),
+    "noisy-x-minus90": expm(+1j * np.pi / 4 * np.array([[0, 1],
+                                                        [1, 0]])),
+}
+
+TWO_Q = {
+    "noisy-cz": np.diag([1, 1, 1, -1]),
+}
+
+
+def noisy_instruction(instruction):
+    if isinstance(instruction, Pragma):
+        return instruction
+
+    if instruction.name == 'RZ':
+        return instruction
+
+    if instruction.name == 'I':
+        return instruction
+
+    if instruction.name == 'RX':
+        assert len(instruction.params) == 1
+        assert len(instruction.qubits) == 1
+        param_str = format_parameter(instruction.params[0])
```

Comment:
This is equivalent to checking if the params[0] == pi/2 (ie. do float comparison vs string)